### ISSUE TITLE
Getting only audio to file/AssetReader ?

### ISSUE THREAD

**Issue by chnbr** [alexeichhorn/YouTubeKit#61]
I need to extract the audio part and save it to a file. Currently my code works like this:

```
// get the stream
let stream = try await YouTube(url: inUrl).streams
  .filterAudioOnly()
  .filter { $0.fileExtension == .m4a }
  .highestAudioBitrateStream()

// download stream url to locale file
(localeUrl, _) = try await URLSession.shared.download(from: stream?.url, delegate: self)

// copy downloaded file to mpeg4audio conforming file
let copyURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0].appendingPathComponent("tmpfile", conformingTo: .mpeg4Audio)
try? FileManager.default.removeItem(at: copyURL)
try FileManager.default.moveItem(at: localeUrl, to: copyURL)

// use AssetReader to read it
let asset = AVURLAsset(url: copyURL)
let assetReader = try AVAssetReader(asset: asset)

```

Now my problem is that I am restricting to m4a audio files which I don't want to. However, if I remove the filter modifier I can get audio which is not coded in MPEG4 audio. Obviously I can't copy such a file with "conformingTo: .mpeg4audio". 

How can I download ALL audio coding types and feed this into AssetReader ? (I think AssetReader needs a correct file extension to work properly. How do I get the audio coding file type of a correct filetype extension for the stream ?)

Thanks !

**Comment by alexeichhorn** [comment]
I usually do it via `.filter { $0.isNativelyPlayable }`, which ensures AVAssetReader can handle it. 
You can use the `.fileExtension.rawValue` property to get the file extension for the cache file. 
But maybe it's a good idea to first be sure that the AVAssetReader actually requires a file extension - would make it easier without of course.

**Comment by alexeichhorn** [comment]
Yes, usually you see `m4a` and `webm` audio files. For example, for video `2lAe1cqCOXo`, you get:
```swift
print(streams.filter { $0.includesAudioTrack }.map { $0.fileExtension })
// → [YouTubeKit.FileExtension.m4a, YouTubeKit.FileExtension.m4a, YouTubeKit.FileExtension.mp4, YouTubeKit.FileExtension.webm, YouTubeKit.FileExtension.webm, YouTubeKit.FileExtension.webm]
```

**Comment by chnbr** [comment]
Hi,

currently with the code segment:

```
let stream = try await YouTube(url: inUrl).streams
    .filterAudioOnly()
    .filter { $0.isNativelyPlayable }
    .highestAudioBitrateStream()
```

I get an exception and find those logs. 

Is YTKit currently broken due to a change at YT ?

> inURL: https://www.youtube.com/watch?v=u_LaF1L-EEs 
> Skipping player response from client 1. Got player response for nil instead of u_LaF1L-EEs
> applying descrambler
> finding initial function name
> Failed to decode object from given start point: Die Daten konnten nicht geöffnet werden, da sie nicht das korrekte Format haben.
> pattern (ytplayer\.config\s*=\s*) failed: Der Vorgang konnte nicht abgeschlossen werden. (YouTubeKit.YouTubeKitError-Fehler 1.)
> -[RTIInputSystemClient remoteTextInputSessionWithID:performInputOperation:]  perform input operation requires a valid sessionID. inputModality = Keyboard, inputOperation = <null selector>, customInfoType = UIEmojiSearchOperations
> -[RTIInputSystemClient remoteTextInputSessionWithID:performInputOperation:]  perform input operation requires a valid sessionID. inputModality = Keyboard, inputOperation = <null selector>, customInfoType = UIEmojiSearchOperations
> applying descrambler
> finding initial function name
> finished regex search, matched \b[a-zA-Z0-9]+\s*&&\s*[a-zA-Z0-9]+\.set\([^,]+\s*,\s*([a-zA-Z0-9$]+)\(
> getting transform plan
> finding initial function name
> Failed to decode object from given start point: Die Daten konnten nicht geöffnet werden, da sie nicht das korrekte Format haben.
> pattern (ytplayer\.config\s*=\s*) failed: Der Vorgang konnte nicht abgeschlossen werden. (YouTubeKit.YouTubeKitError-Fehler 1.)
> finding initial function name
> finished regex search, matched \b[a-zA-Z0-9]+\s*&&\s*[a-zA-Z0-9]+\.set\([^,]+\s*,\s*([a-zA-Z0-9$]+)\(
> getting transform plan
>

**Comment by chnbr** [comment]
The problem still persists. Any comments ?

Thanks

**Comment by chnbr** [comment]
Hi Alex,

I have updated to 0.2.5 but still I have problems with:

          let stream = try await YouTube(url: inUrl).streams
            .filterAudioOnly()
            .filter { $0.isNativelyPlayable }
            .highestAudioBitrateStream()

It worked a few days ago and now it fails again. I get the output:

from my logs: inURL: https://www.youtube.com/watch?v=u_LaF1L-EEs 

finding initial function name
pattern (ytplayer\.config\s*=\s*) failed: Der Vorgang konnte nicht abgeschlossen werden. (YouTubeKit.YouTubeKitError-Fehler 1.)
applying descrambler
applying descrambler
finding initial function name
getting transform plan
finding initial function name
pattern (ytplayer\.config\s*=\s*) failed: Der Vorgang konnte nicht abgeschlossen werden. (YouTubeKit.YouTubeKitError-Fehler 1.)
finding initial function name
getting transform plan


Do you have any ideas what could be wrong here ? Is YT changing their API so frequently ?

Thanks and regards
Christian
